src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -268,52 +269,88 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
             var projects = (await solutionManager.GetNuGetProjectsAsync()).OfType<IDependencyGraphProject>().ToList();
             var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
+                await actionBlock.SendAsync(new ProjectRestoreSpec()
```

Comment:
Is the `context` object thread-safe? Or rather, does `project.GetPackageSpecsAndAdditionalMessagesAsync` use `context` in a thread-safe way?

If the action block is changed to `ActionBlock<IDependencyGraphProject>`, then you can send the porject instances to dataflow, and do all the `GetPackageSpecsAndAdditionalMessagesAsync` calls in parallel as well.

Reply:
@zivkan  Here it looks `GetPackageSpecsAndAdditionalMessagesAsync` doesn't use sync locks for `context` before mutating it.
[MSBuildNuGetProject.cs#L699](https://github.com/NuGet/NuGet.Client/blob/78adad7172e2e6eac269454583e9c0262fdefa2d/src/NuGet.Core/NuGet.PackageManagement/Projects/MSBuildNuGetProject.cs#L699)
[ProjectJsonNuGetProject.cs#L254](https://github.com/NuGet/NuGet.Client/blob/78adad7172e2e6eac269454583e9c0262fdefa2d/src/NuGet.Core/NuGet.PackageManagement/Projects/ProjectJsonNuGetProject.cs#L254)

Because here `PackageSpecCache` is normal dictionary so I don't believe it's thread safe.